subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy

```diff
@@ -815,6 +815,128 @@ class JavaCompileIntegrationTest extends AbstractIntegrationSpec {
         !file("build/classes/java/main/ignored/io/ignored/IgnoredExample.class").exists()
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK16_OR_LATER)
+    def "class snapshotting ignores changes to a record implementation"() {
+        given:
+        settingsFile << "include 'a', 'b'"
+
+
+        file('a/build.gradle') << """
+            apply plugin: 'java'
+        """
+        file('a/src/main/java/Foo.java') << 'record Foo(String property) {}'
+
+        file('b/build.gradle') << '''
+            apply plugin: 'java'
+
+            dependencies {
+                implementation project(':a')
+            }
+        '''
+        file('b/src/main/java/Bar.java') << 'class Bar { public void useFoo(Foo foo) { } }'
+
+        when:
+        // Run with --debug to ensure that class snapshotting didn't fail.
+        succeeds ":b:compileJava", "--debug"
+
+        then:
+        executedAndNotSkipped(":b:compileJava")
+
+        when:
+        // Change internal implementation, but not API.
+        file('a/src/main/java/Foo.java').delete()
+        file('a/src/main/java/Foo.java') << '''
+record Foo(String property) {
+    public String property() {
+        return property + "!";
+    }
+}
+'''
+
+        succeeds ":b:compileJava"
+
+        then:
+        skipped(":b:compileJava")
+
+        when:
+        // Add new property, breaks API.
+        file('a/src/main/java/Foo.java').delete()
+        file('a/src/main/java/Foo.java') << '''
+record Foo(String property, int newProperty) {
+    public String property() {
+        return property + "!";
+    }
+}
+'''
+
+        succeeds ":b:compileJava"
+
+        then:
+        executedAndNotSkipped(":b:compileJava")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK17_OR_LATER)
```

Comment:
Aren't sealed modifiers a JDK15 feature?  Or is there some other compatibility issue here I'm not aware of?

Reply:
They are a JDK 15 _preview_ feature, but only stabilized in 17 [[source]](https://openjdk.java.net/projects/jdk/17/). I can add special-casing to compile with `--preview` if we want, when needed, but I think the tests are sufficient as they are.

Reply:
Nope, that makes sense.  I wasn't aware these were only preview.